File: source/modulo_new_core/src/translators/parameter_translators.cpp
#include "modulo_new_core/translators/parameter_translators.hpp"

#include <state_representation/space/cartesian/CartesianState.hpp>
#include <state_representation/space/cartesian/CartesianPose.hpp>
#include <state_representation/space/joint/JointState.hpp>
#include <state_representation/space/joint/JointPositions.hpp>

#include "modulo_new_core/exceptions/IncompatibleParameterException.hpp"

#include <clproto.h>

using namespace state_representation;

namespace modulo_new_core::translators {

void copy_parameter_value(
    const std::shared_ptr<ParameterInterface>& source_parameter, const std::shared_ptr<ParameterInterface>& parameter
) {
  if (source_parameter->get_parameter_type() != parameter->get_parameter_type()) {
    throw exceptions::IncompatibleParameterException(
        "Source parameter " + source_parameter->get_name()
            + " to be copied does not have the same type as destination parameter " + parameter->get_name());
  }
  switch (source_parameter->get_parameter_type()) {
    case ParameterType::BOOL:
      parameter->set_parameter_value(source_parameter->get_parameter_value<bool>());
      return;
    case ParameterType::BOOL_ARRAY:
      parameter->set_parameter_value(source_parameter->get_parameter_value<std::vector<bool>>());
      return;
    case ParameterType::INT:
      parameter->set_parameter_value(source_parameter->get_parameter_value<int>());
      return;
    case ParameterType::INT_ARRAY:
      parameter->set_parameter_value(source_parameter->get_parameter_value<std::vector<int>>());
      return;
    case ParameterType::DOUBLE:
      parameter->set_parameter_value(source_parameter->get_parameter_value<double>());
      return;
    case ParameterType::DOUBLE_ARRAY:
      parameter->set_parameter_value(source_parameter->get_parameter_value<std::vector<double>>());
      return;
    case ParameterType::STRING:
      parameter->set_parameter_value(source_parameter->get_parameter_value<std::string>());
      return;
    case ParameterType::STRING_ARRAY:
      parameter->set_parameter_value(source_parameter->get_parameter_value<std::vector<std::string>>());
      return;
    case ParameterType::VECTOR:
      parameter->set_parameter_value(source_parameter->get_parameter_value<Eigen::VectorXd>());
      return;
    case ParameterType::MATRIX:
      parameter->set_parameter_value(source_parameter->get_parameter_value<Eigen::MatrixXd>());
      return;
    case ParameterType::STATE:
      if (source_parameter->get_parameter_state_type() != parameter->get_parameter_state_type()) {
        throw exceptions::IncompatibleParameterException(
            "Source parameter " + source_parameter->get_name()
                + " to be copied does not have the same parameter state type as destination parameter "
                + parameter->get_name());
      }
      switch (source_parameter->get_parameter_state_type()) {
        case state_representation::StateType::CARTESIAN_STATE:
          parameter->set_parameter_value(source_parameter->get_parameter_value<CartesianState>());
          return;
        case state_representation::StateType::CARTESIAN_POSE:
          parameter->set_parameter_value(source_parameter->get_parameter_value<CartesianPose>());
          return;
        case state_representation::StateType::JOINT_STATE:
          parameter->set_parameter_value(source_parameter->get_parameter_value<JointState>());
          return;
        case state_representation::StateType::JOINT_POSITIONS:
          parameter->set_parameter_value(source_parameter->get_parameter_value<JointPositions>());
          return;
        default:
          break;
      }
      break;
    default:
      break;
  }
  throw InvalidParameterException(
      "Could not copy the value from source parameter " + source_parameter->get_name() + " into parameter "
          + parameter->get_name());
}

rclcpp::Parameter write_parameter(const std::shared_ptr<state_representation::ParameterInterface>& parameter) {
  switch (parameter->get_parameter_type()) {
    case ParameterType::BOOL:
      return {parameter->get_name(), parameter->get_parameter_value<bool>()};
    case ParameterType::BOOL_ARRAY:
      return {parameter->get_name(), parameter->get_parameter_value<std::vector<bool>>()};
    case ParameterType::INT:
      return {parameter->get_name(), parameter->get_parameter_value<int>()};
    case ParameterType::INT_ARRAY:
      return {parameter->get_name(), parameter->get_parameter_value<std::vector<int>>()};
    case ParameterType::DOUBLE:
      return {parameter->get_name(), parameter->get_parameter_value<double>()};
    case ParameterType::DOUBLE_ARRAY:
      return {parameter->get_name(), parameter->get_parameter_value<std::vector<double>>()};
    case ParameterType::STRING:
      return {parameter->get_name(), parameter->get_parameter_value<std::string>()};
    case ParameterType::STRING_ARRAY:
      return {parameter->get_name(), parameter->get_parameter_value<std::vector<std::string>>()};
    case ParameterType::STATE: {
      switch (parameter->get_parameter_state_type()) {
        case StateType::CARTESIAN_STATE:
          return {parameter->get_name(), clproto::to_json(parameter->get_parameter_value<CartesianState>())};
        case StateType::CARTESIAN_POSE:
          return {parameter->get_name(), clproto::to_json(parameter->get_parameter_value<CartesianPose>())};
        case StateType::JOINT_STATE:
          return {parameter->get_name(), clproto::to_json(parameter->get_parameter_value<JointState>())};
        case StateType::JOINT_POSITIONS:
          return {parameter->get_name(), clproto::to_json(parameter->get_parameter_value<JointPositions>())};
        default:
          break;
      }
      break;
    }
    case ParameterType::VECTOR: {
      auto eigen_vector = parameter->get_parameter_value<Eigen::VectorXd>();
      std::vector<double> vec(eigen_vector.data(), eigen_vector.data() + eigen_vector.size());
      return {parameter->get_name(), vec};
    }
    case ParameterType::MATRIX: {
      auto eigen_matrix = parameter->get_parameter_value<Eigen::MatrixXd>();
      std::vector<double> vec(eigen_matrix.data(), eigen_matrix.data() + eigen_matrix.size());
      return {parameter->get_name(), vec};
    }
    default:
      break;
  }
  throw InvalidParameterException("Parameter " + parameter->get_name() + " could not be written!");
}

std::shared_ptr<state_representation::ParameterInterface> read_parameter(const rclcpp::Parameter& parameter) {
  switch (parameter.get_type()) {
    case rclcpp::PARAMETER_BOOL:
      return make_shared_parameter(parameter.get_name(), parameter.as_bool());
    case rclcpp::PARAMETER_BOOL_ARRAY:
      return make_shared_parameter(parameter.get_name(), parameter.as_bool_array());
    case rclcpp::PARAMETER_INTEGER:
      return make_shared_parameter(parameter.get_name(), static_cast<int>(parameter.as_int()));
    case rclcpp::PARAMETER_INTEGER_ARRAY: {
      auto array = parameter.as_integer_array();
      std::vector<int> int_array(array.begin(), array.end());
      return make_shared_parameter(parameter.get_name(), int_array);
    }
    case rclcpp::PARAMETER_DOUBLE:
      return make_shared_parameter(parameter.get_name(), parameter.as_double());
    case rclcpp::PARAMETER_DOUBLE_ARRAY:
      return make_shared_parameter(parameter.get_name(), parameter.as_double_array());
    case rclcpp::PARAMETER_STRING_ARRAY:
      return make_shared_parameter<std::vector<std::string>>(parameter.get_name(), parameter.as_string_array());
    case rclcpp::PARAMETER_STRING: {
      // TODO: consider dedicated clproto::decode<std::shared_ptr<ParameterInterface>>(msg) specialization in library
      std::string encoding;
      try {
        encoding = clproto::from_json(parameter.as_string());
      } catch (const clproto::JsonParsingException&) {}
      if (!clproto::is_valid(encoding)) {
        return make_shared_parameter<std::string>(parameter.get_name(), parameter.as_string());
      }
      switch (clproto::check_message_type(encoding)) {
        case clproto::CARTESIAN_STATE_MESSAGE:
          return make_shared_parameter<CartesianState>(parameter.get_name(), clproto::decode<CartesianState>(encoding));
        case clproto::CARTESIAN_POSE_MESSAGE:
          return make_shared_parameter<CartesianPose>(parameter.get_name(), clproto::decode<CartesianPose>(encoding));
        case clproto::JOINT_STATE_MESSAGE:
          return make_shared_parameter<JointState>(parameter.get_name(), clproto::decode<JointState>(encoding));
        case clproto::JOINT_POSITIONS_MESSAGE:
          return make_shared_parameter<JointPositions>(parameter.get_name(), clproto::decode<JointPositions>(encoding));
        default:
          throw InvalidParameterException(
              "Parameter " + parameter.get_name() + " has an unsupported encoded message type");
      }
    }
    case rclcpp::PARAMETER_BYTE_ARRAY:
      // TODO: try clproto decode, re-use logic from above
      throw InvalidParameterException("Parameter byte arrays are not currently supported.");
    default:
      break;
  }
  throw InvalidParameterException("Parameter " + parameter.get_name() + " could not be read!");
}

std::shared_ptr<state_representation::ParameterInterface> read_parameter_const(
    const rclcpp::Parameter& ros_parameter,
    const std::shared_ptr<const state_representation::ParameterInterface>& parameter
) {
  if (ros_parameter.get_name() != parameter->get_name()) {
    throw exceptions::IncompatibleParameterException(
        "The ROS parameter " + ros_parameter.get_name()
            + " to be read does not have the same name as the reference parameter " + parameter->get_name());
  }
  auto new_parameter = read_parameter(ros_parameter);
  if (new_parameter->get_parameter_type() == parameter->get_parameter_type()) {
    return new_parameter;
  }
  switch (new_parameter->get_parameter_type()) {
    case ParameterType::DOUBLE_ARRAY: {
      auto value = new_parameter->get_parameter_value<std::vector<double>>();
      switch (parameter->get_parameter_type()) {
        case ParameterType::VECTOR: {
          Eigen::VectorXd vector = Eigen::Map<Eigen::VectorXd>(value.data(), static_cast<Eigen::Index>(value.size()));
          new_parameter = make_shared_parameter(parameter->get_name(), vector);
          break;
        }
        case ParameterType::MATRIX: {
          /* TODO: get_parameter_value must be const
          auto matrix = parameter->get_parameter_value<Eigen::MatrixXd>();
          if (static_cast<std::size_t>(matrix.size()) != value.size()) {
            throw exceptions::IncompatibleParameterException(
                "The ROS parameter " + ros_parameter.get_name() + " with type double array has size "
                    + std::to_string(value.size()) + " while the reference parameter matrix " + parameter->get_name()
                    + " has size " + std::to_string(matrix.size()));
          }
          matrix = Eigen::Map<Eigen::MatrixXd>(value.data(), matrix.rows(), matrix.cols());
          new_parameter = make_shared_parameter(parameter->get_name(), matrix)
          break;
           */
        }
        default:
          throw exceptions::IncompatibleParameterException(
              "The ROS parameter " + ros_parameter.get_name()
                  + " with type double array cannot be interpreted by reference parameter " + parameter->get_name()
                  + " (type code " + std::to_string(static_cast<int>(parameter->get_parameter_type())) + ")");
          break;
      }
      break;
    }
    default:
      throw state_representation::exceptions::InvalidParameterException(
          "Something went wrong while reading parameter " + parameter->get_name());
      break;
  }
  return new_parameter;
}

void read_parameter(
    const rclcpp::Parameter& ros_parameter, const std::shared_ptr<ParameterInterface>& parameter
) {
  auto new_parameter = read_parameter_const(ros_parameter, parameter);
  copy_parameter_value(new_parameter, parameter);
}

}

File: source/modulo_new_core/modulo_new_core/translators/parameter_translators.py
from rclpy import Parameter
import state_representation as sr
import clproto
import numpy as np
from typing import Union


def write_parameter(parameter: sr.Parameter) -> Parameter:
    if parameter.get_parameter_type() == sr.ParameterType.BOOL or \
            parameter.get_parameter_type() == sr.ParameterType.BOOL_ARRAY or \
            parameter.get_parameter_type() == sr.ParameterType.INT or \
            parameter.get_parameter_type() == sr.ParameterType.INT_ARRAY or \
            parameter.get_parameter_type() == sr.ParameterType.DOUBLE or \
            parameter.get_parameter_type() == sr.ParameterType.DOUBLE_ARRAY or \
            parameter.get_parameter_type() == sr.ParameterType.STRING or \
            parameter.get_parameter_type() == sr.ParameterType.STRING_ARRAY:
        return Parameter(parameter.get_name(), value=parameter.get_value())
    elif parameter.get_parameter_type() == sr.ParameterType.STATE:
        if parameter.get_parameter_state_type() == sr.StateType.CARTESIAN_STATE:
            return Parameter(parameter.get_name(), Parameter.Type.STRING, clproto.to_json(
                clproto.encode(parameter.get_value(), clproto.MessageType.CARTESIAN_STATE_MESSAGE)))
        elif parameter.get_parameter_state_type() == sr.StateType.CARTESIAN_POSE:
            return Parameter(parameter.get_name(), Parameter.Type.STRING, clproto.to_json(
                clproto.encode(parameter.get_value(), clproto.MessageType.CARTESIAN_POSE_MESSAGE)))
        elif parameter.get_parameter_state_type() == sr.StateType.JOINT_STATE:
            return Parameter(parameter.get_name(), Parameter.Type.STRING, clproto.to_json(
                clproto.encode(parameter.get_value(), clproto.MessageType.JOINT_STATE_MESSAGE)))
        if parameter.get_parameter_state_type() == sr.StateType.JOINT_POSITIONS:
            return Parameter(parameter.get_name(), Parameter.Type.STRING, clproto.to_json(
                clproto.encode(parameter.get_value(), clproto.MessageType.JOINT_POSITIONS_MESSAGE)))
        else:
            raise RuntimeError(f"Parameter {parameter.get_name()} could not be written!")
    elif parameter.get_parameter_type() == sr.ParameterType.VECTOR or \
            parameter.get_parameter_type() == sr.ParameterType.MATRIX:
        return Parameter(parameter.get_name(), value=parameter.get_value().flatten().tolist())
    else:
        raise RuntimeError(f"Parameter {parameter.get_name()} could not be written!")


def copy_parameter_value(source_parameter: sr.Parameter, parameter: sr.Parameter):
    if source_parameter.get_parameter_type() != parameter.get_parameter_type():
        raise RuntimeError(f"Source parameter {source_parameter.get_name()} to be copied does not have the same type "
                           f"as destination parameter {parameter.get_name()}")
    if source_parameter.get_parameter_type() == sr.ParameterType.BOOL or \
            source_parameter.get_parameter_type() == sr.ParameterType.BOOL_ARRAY or \
            source_parameter.get_parameter_type() == sr.ParameterType.INT or \
            source_parameter.get_parameter_type() == sr.ParameterType.INT_ARRAY or \
            source_parameter.get_parameter_type() == sr.ParameterType.DOUBLE or \
            source_parameter.get_parameter_type() == sr.ParameterType.DOUBLE_ARRAY or \
            source_parameter.get_parameter_type() == sr.ParameterType.STRING or \
            source_parameter.get_parameter_type() == sr.ParameterType.STRING_ARRAY or \
            source_parameter.get_parameter_type() == sr.ParameterType.VECTOR or \
            source_parameter.get_parameter_type() == sr.ParameterType.MATRIX:
        parameter.set_value(source_parameter.get_value())
    elif source_parameter.get_parameter_type() == sr.ParameterType.STATE:
        if source_parameter.get_parameter_state_type() != parameter.get_parameter_state_type():
            raise RuntimeError(
                f"Source parameter {source_parameter.get_name()} to be copied does not have the same state type "
                f"as destination parameter {parameter.get_name()}")
        if source_parameter.get_parameter_state_type() == sr.StateType.CARTESIAN_STATE or \
                source_parameter.get_parameter_state_type() == sr.StateType.CARTESIAN_POSE or \
                source_parameter.get_parameter_state_type() == sr.StateType.JOINT_STATE or \
                source_parameter.get_parameter_state_type() == sr.StateType.JOINT_POSITIONS:
            parameter.set_value(source_parameter.get_value())
        else:
            raise RuntimeError(f"Could not copy the value from source parameter {source_parameter.get_name()} "
                               f"into parameter {parameter.get_name()}")
    else:
        raise RuntimeError(f"Could not copy the value from source parameter {source_parameter.get_name()} "
                           f"into parameter {parameter.get_name()}")


def read_parameter(ros_parameter: Parameter, parameter: Union[None, sr.Parameter] = None) -> sr.Parameter:
    def read_new_parameter(ros_param: Parameter) -> sr.Parameter:
        if ros_param.type_ == Parameter.Type.BOOL:
            return sr.Parameter(ros_param.name, ros_param.get_parameter_value().bool_value, sr.ParameterType.BOOL)
        elif ros_param.type_ == Parameter.Type.BOOL_ARRAY:
            return sr.Parameter(ros_param.name, ros_param.get_parameter_value().bool_array_value,
                                sr.ParameterType.BOOL_ARRAY)
        elif ros_param.type_ == Parameter.Type.INTEGER:
            return sr.Parameter(ros_param.name, ros_param.get_parameter_value().integer_value, sr.ParameterType.INT)
        elif ros_param.type_ == Parameter.Type.INTEGER_ARRAY:
            return sr.Parameter(ros_param.name, ros_param.get_parameter_value().integer_array_value,
                                sr.ParameterType.INT_ARRAY)
        elif ros_param.type_ == Parameter.Type.DOUBLE:
            return sr.Parameter(ros_param.name, ros_param.get_parameter_value().double_value, sr.ParameterType.DOUBLE)
        elif ros_param.type_ == Parameter.Type.DOUBLE_ARRAY:
            return sr.Parameter(ros_param.name, ros_param.get_parameter_value().double_array_value,
                                sr.ParameterType.DOUBLE_ARRAY)
        elif ros_param.type_ == Parameter.Type.STRING_ARRAY:
            return sr.Parameter(ros_param.name, ros_param.get_parameter_value().string_array_value,
                                sr.ParameterType.STRING_ARRAY)
        elif ros_param.type_ == Parameter.Type.STRING:
            encoding = ""
            try:
                encoding = clproto.from_json(ros_param.get_parameter_value().string_value)
            except Exception:
                if not clproto.is_valid(encoding):
                    return sr.Parameter(ros_param.name, ros_param.get_parameter_value().string_value,
                                        sr.ParameterType.STRING)
            message_type = clproto.check_message_type(encoding)
            if message_type == clproto.MessageType.CARTESIAN_STATE_MESSAGE:
                return sr.Parameter(ros_param.name, clproto.decode(encoding), sr.ParameterType.STATE,
                                    sr.StateType.CARTESIAN_STATE)
            elif message_type == clproto.MessageType.CARTESIAN_POSE_MESSAGE:
                return sr.Parameter(ros_param.name, clproto.decode(encoding), sr.ParameterType.STATE,
                                    sr.StateType.CARTESIAN_POSE)
            elif message_type == clproto.MessageType.JOINT_STATE_MESSAGE:
                return sr.Parameter(ros_param.name, clproto.decode(encoding), sr.ParameterType.STATE,
                                    sr.StateType.JOINT_STATE)
            elif message_type == clproto.MessageType.JOINT_POSITIONS_MESSAGE:
                return sr.Parameter(ros_param.name, clproto.decode(encoding), sr.ParameterType.STATE,
                                    sr.StateType.JOINT_POSITIONS)
            else:
                raise RuntimeError(f"Parameter {ros_param.name} has an unsupported encoded message type!")
        elif ros_param.type_ == Parameter.Type.BYTE_ARRAY:
            raise RuntimeError("Parameter byte arrays are not currently supported.")
        else:
            raise RuntimeError(f"Parameter {ros_param.name} could not be read!")

    new_parameter = read_new_parameter(ros_parameter)
    if parameter is None:
        return new_parameter
    else:
        copy_parameter_value(new_parameter, parameter)
        return parameter


def read_parameter_const(ros_parameter: Parameter, parameter: sr.Parameter) -> sr.Parameter:
    if ros_parameter.name != parameter.get_name():
        raise RuntimeError(
            f"The ROS parameter {ros_parameter.name} to be read does not have the same name as the reference parameter {parameter.get_name()}")
    new_parameter = read_parameter(ros_parameter)
    if new_parameter.get_parameter_type() == parameter.get_parameter_type():
        return new_parameter
    elif new_parameter.get_parameter_type() == sr.ParameterType.DOUBLE_ARRAY:
        value = new_parameter.get_value()
        if parameter.get_parameter_type() == sr.ParameterType.VECTOR:
            vector = np.array(value)
            return sr.Parameter(parameter.get_name(), vector, sr.ParameterType.VECTOR)
        elif parameter.get_parameter_type() == sr.ParameterType.MATRIX:
            matrix = parameter.get_value()
            if matrix.size != len(value):
                raise RuntimeError(f"The ROS parameter {ros_parameter.name} with type double array has size "
                                   f"{len(value)} while the reference parameter matrix {parameter.get_name()} "
                                   f"has size {len(matrix)}")
            matrix = np.array(value).reshape(matrix.shape)
            return sr.Parameter(parameter.get_name(), matrix, sr.ParameterType.MATRIX)
        else:
            raise RuntimeError(f"The ROS parameter {ros_parameter.name} with type double array cannot be interpreted "
                               f"by reference parameter {parameter.get_name()} (type code "
                               f"{parameter.get_parameter_type()}")
    else:
        raise RuntimeError(f"Something went wrong while reading parameter {parameter.get_name()}")
